webview-ui/src/ConnectAcrToCluster/state/dataLoading.ts

```diff
@@ -0,0 +1,139 @@
+    }
+
+    if (isNotLoaded(subscriptionData.acrs)) {
+        const key = { subscriptionId: subscriptionData.subscription.subscriptionId };
+        vscode.postGetAcrsRequest(key);
+        updates.push((e) => e.onSetAcrsLoading(key));
+    }
+
+    return lazyMap(subscriptionData.acrs, (data) => data.map((a) => a.acr));
+}
+
+export function ensureClustersLoaded(
+    referenceData: AzureReferenceData,
+    subscription: Subscription | null,
+    updates: EventHandlerFunc[],
+): Lazy<Cluster[]> {
+    const subscriptionData = getSubscriptionReferenceData(referenceData, subscription);
+    if (subscriptionData === null) {
+        return newNotLoaded();
+    }
+
+    if (isNotLoaded(subscriptionData.clusters)) {
+        const key = { subscriptionId: subscriptionData.subscription.subscriptionId };
+        vscode.postGetClustersRequest(key);
+        updates.push((e) => e.onSetClustersLoading(key));
+    }
+
+    return subscriptionData.clusters;
+}
+
+export function ensureAcrRoleAssignmentLoaded(
+    referenceData: AzureReferenceData,
+    subscription: Subscription | null,
+    acrKey: AcrKey | null,
+    clusterKey: ClusterKey | null,
+    updates: EventHandlerFunc[],
+): Lazy<AcrRoleState> {
+    const acrData = getAcrReferenceData(referenceData, subscription, acrKey);
+    if (acrData === null) {
+        return newNotLoaded();
+    }
+
+    if (clusterKey === null) {
+        return newNotLoaded();
+    }
+
+    const assignedRoleDefinitions: Lazy<AcrRoleState> =
+        acrData.assignedRoleDefinitions[createClusterId(clusterKey)] || newNotLoaded();
+
+    if (isNotLoaded(assignedRoleDefinitions)) {
+        const acrKey: AcrKey = {
+            subscriptionId: acrData.acr.subscriptionId,
+            resourceGroup: acrData.acr.resourceGroup,
+            acrName: acrData.acr.acrName,
+        };
+
+        vscode.postGetAcrRoleAssignmentRequest({ acrKey, clusterKey });
+        updates.push((e) => e.onSetAcrRoleAssignmentLoading({ acrKey, clusterKey }));
+    }
+
+    return assignedRoleDefinitions;
+}
+
+function getSubscriptionReferenceData(
+    referenceData: AzureReferenceData,
+    subscription: Subscription | null,
+): SubscriptionReferenceData | null {
+    if (!isLoaded(referenceData.subscriptions) || subscription === null) {
+        return null;
+    }
+
+    return getOrThrow(
+        referenceData.subscriptions.value,
+        (s) => s.subscription.subscriptionId === subscription.subscriptionId,
+        `${subscription.subscriptionId} (${subscription.name}) not found`,
+    );
+}
+
+function getAcrReferenceData(
+    referenceData: AzureReferenceData,
+    subscription: Subscription | null,
+    acr: Acr | null,
+): AcrReferenceData | null {
+    const subscriptionData = getSubscriptionReferenceData(referenceData, subscription);
+    if (subscriptionData === null || !isLoaded(subscriptionData.acrs)) {
+        return null;
+    }
+
+    if (subscription === null || acr === null) {
+        return null;
+    }
```

Comment:
No strong preference here. Part of me likes the separate code paths because it helps to distinguish between "returning because we can't get reference data" and "returning because other arguments are not set".

But it's really not important - I'll change it.